wgpu-core/src/device/life.rs

```diff
@@ -14,111 +14,81 @@ use crate::{
         TextureViewId,
     },
     pipeline::{ComputePipeline, RenderPipeline},
-    resource::{
-        self, Buffer, QuerySet, Resource, ResourceType, Sampler, StagingBuffer, Texture,
-        TextureView,
-    },
+    resource::{self, Buffer, QuerySet, Resource, Sampler, StagingBuffer, Texture, TextureView},
     track::{ResourceTracker, Tracker},
     FastHashMap, SubmissionIndex,
 };
 use smallvec::SmallVec;
 
 use parking_lot::Mutex;
+use std::sync::Arc;
 use thiserror::Error;
-use wgt::WasmNotSendSync;
-
-use std::{any::Any, sync::Arc};
-
-pub(crate) trait ResourceMap: Any + WasmNotSendSync {
-    fn as_any(&self) -> &dyn Any;
-    fn as_any_mut(&mut self) -> &mut dyn Any;
-    fn clear_map(&mut self);
-    fn extend_map(&mut self, maps: &mut ResourceMaps);
-}
-
-impl<Id, R> ResourceMap for FastHashMap<Id, Arc<R>>
-where
-    Id: id::TypedId,
-    R: Resource<Id>,
-{
-    fn as_any(&self) -> &dyn Any {
-        self
-    }
-    fn as_any_mut(&mut self) -> &mut dyn Any {
-        self
-    }
-    fn clear_map(&mut self) {
-        self.clear()
-    }
-    fn extend_map(&mut self, r: &mut ResourceMaps) {
-        if let Some(other) = r.maps.get_mut(R::TYPE) {
-            if let Some(other) = other.as_any_mut().downcast_mut::<Self>() {
-                self.extend(other.drain());
-            }
-        }
-    }
-}
 
 /// A struct that keeps lists of resources that are no longer needed by the user.
 #[derive(Default)]
-pub(crate) struct ResourceMaps {
-    pub(crate) maps: FastHashMap<ResourceType, Box<dyn ResourceMap>>,
+pub(crate) struct ResourceMaps<A: HalApi> {
+    pub(crate) buffers: FastHashMap<BufferId, Arc<Buffer<A>>>,
+    pub(crate) staging_buffers: FastHashMap<StagingBufferId, Arc<StagingBuffer<A>>>,
+    pub(crate) textures: FastHashMap<TextureId, Arc<Texture<A>>>,
+    pub(crate) texture_views: FastHashMap<TextureViewId, Arc<TextureView<A>>>,
+    pub(crate) samplers: FastHashMap<SamplerId, Arc<Sampler<A>>>,
+    pub(crate) bind_groups: FastHashMap<BindGroupId, Arc<BindGroup<A>>>,
+    pub(crate) bind_group_layouts: FastHashMap<BindGroupLayoutId, Arc<BindGroupLayout<A>>>,
+    pub(crate) render_pipelines: FastHashMap<RenderPipelineId, Arc<RenderPipeline<A>>>,
+    pub(crate) compute_pipelines: FastHashMap<ComputePipelineId, Arc<ComputePipeline<A>>>,
+    pub(crate) pipeline_layouts: FastHashMap<PipelineLayoutId, Arc<PipelineLayout<A>>>,
+    pub(crate) render_bundles: FastHashMap<RenderBundleId, Arc<RenderBundle<A>>>,
+    pub(crate) query_sets: FastHashMap<QuerySetId, Arc<QuerySet<A>>>,
```

Comment:
nit: `pub(crate)` clutter. You _could_ just make all of this regular `pub` since `life` is dragged in as `mod life` (not `pub mod life`)